src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -40,6 +101,283 @@ public override bool CanConvert(Type typeToConvert)
         /// <returns>The value that was converted.</returns>
         public abstract T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options);
 
+        internal bool TryRead(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options, ref ReadStack state, out T value)
+        {
+            if (ClassType == ClassType.Value)
+            {
+                // A value converter should never be within a continuation.
+                Debug.Assert(!state.IsContinuation);
+
+                // For perf and converter simplicity, handle null here instead of forwarding to the converter.
+                if (reader.TokenType == JsonTokenType.Null && !HandleNullValue)
+                {
+                    value = default!;
+                    return true;
+                }
+
+#if !DEBUG
+                // For performance, only perform validation on internal converters on debug builds.
+                if (IsInternalConverter)
+                {
+                    value = Read(ref reader, typeToConvert, options);
+                }
+                else
+#endif
+                {
+                    JsonTokenType originalPropertyTokenType = reader.TokenType;
+                    int originalPropertyDepth = reader.CurrentDepth;
+                    long originalPropertyBytesConsumed = reader.BytesConsumed;
+
+                    value = Read(ref reader, typeToConvert, options);
+                    VerifyRead(
+                        originalPropertyTokenType,
+                        originalPropertyDepth,
+                        originalPropertyBytesConsumed != reader.BytesConsumed,
```

Comment:
`VerifyRead` is called from two locations and one location needs to pass in `true` for that argument.

Previously we didn't have verification for objects and collections, just values (since we didn't have converters for objects and collections).